src/mame/nabu/nabupc.cpp

```diff
@@ -0,0 +1,509 @@
+	save_item(NAME(m_control));
+	save_item(NAME(m_bios_size));
+}
+
+// Machine Reset
+void nabupc_state::machine_reset()
+{
+	m_irq_in_prio = 0xFF;
+	m_int_lines = 0;
+	m_porta = 0;
+	m_portb = 0;
+	m_control = 0;
+	m_bios_size = m_bios_sel->read() == 1 ? 0x2000 : 0x1000;
+	m_leds[0] = 1; // Power LED
+}
+
+
+//**************************************************************************
+//  Control/Status Ports
+//**************************************************************************
+
+// Control Register
+void nabupc_state::control_w(uint8_t data)
+{
+	m_control = data;
+	m_centronics->write_strobe(BIT(m_control, 2));
+	for (int i = 3 ; i < 6 ; ++i) {
+		m_leds[6 - i] = BIT(m_control, i);
+	}
+}
+
+// Status Port
+uint8_t nabupc_state::psg_portb_r()
+{
+	return m_portb;
+}
+
+// Interrupt Enable Port
+void nabupc_state::psg_porta_w(uint8_t data)
+{
+	if (data != m_porta) {
+		m_porta = data;
+		update_irq();
+	}
+}
+
+// Printer Busy State
+void nabupc_state::centronics_busy_handler(uint8_t state)
+{
+	BIT_SET(m_portb, 4, state);
+}
+
+// HCCA Framing Error
+WRITE_LINE_MEMBER(nabupc_state::hcca_fe_w)
+{
+	BIT_SET(m_portb, 5, state);
+}
+
+// HCCA Overrun Error
+WRITE_LINE_MEMBER(nabupc_state::hcca_oe_w)
+{
+	BIT_SET(m_portb, 6, state);
+}
+
+
+//**************************************************************************
+//  Interrupt Handling
+//**************************************************************************
+
+// IRQ For HCCA Receive
+WRITE_LINE_MEMBER(nabupc_state::hcca_dr_w)
+{
+	BIT_SET(m_int_lines, 7, state);
+	update_irq();
+}
+
+// IRQ For HCCA Transmit
+WRITE_LINE_MEMBER(nabupc_state::hcca_tbre_w)
+{
+	BIT_SET(m_int_lines, 6, state);
+	update_irq();
+}
+
+// IRQ For Keyboard
+WRITE_LINE_MEMBER(nabupc_state::rxrdy_w)
+{
+	BIT_SET(m_int_lines, 5, state);
+	update_irq();
+}
+
+// IRQ For TMS9918a
+WRITE_LINE_MEMBER(nabupc_state::vdp_int_w)
+{
+	BIT_SET(m_int_lines, 4, state);
+	update_irq();
+}
+
+// IRQ For Option Slot 1
+WRITE_LINE_MEMBER(nabupc_state::j9_int_w)
+{
+	BIT_SET(m_int_lines, 3, !state);
+	update_irq();
+}
+
+// IRQ For Option Slot 2
+WRITE_LINE_MEMBER(nabupc_state::j10_int_w)
+{
+	BIT_SET(m_int_lines, 2, !state);
+	update_irq();
+}
+
+// IRQ For Option Slot 3
+WRITE_LINE_MEMBER(nabupc_state::j11_int_w)
+{
+	BIT_SET(m_int_lines, 1, !state);
+	update_irq();
+}
+
+// IRQ For Option Slot 4
+WRITE_LINE_MEMBER(nabupc_state::j12_int_w)
+{
+	BIT_SET(m_int_lines, 0, !state);
+	update_irq();
+}
+
+// IRQ Acknowledge
+IRQ_CALLBACK_MEMBER(nabupc_state::int_ack_cb)
+{
+	uint32_t vector = (m_portb) & 0xe;
+	return vector;
+}
+
+// Update CPU IRQ line
+void nabupc_state::update_irq()
+{
+	uint8_t interrupts = ~(m_porta & m_int_lines);
+	m_irq_in_prio = (m_irq_in_prio & 0xff00) | interrupts;
+
+	m_portb &= 0xf0;
+	f9318_out_t out = f9318(static_cast<f9318_in_t>(m_irq_in_prio));
+	m_portb |= ((out & 7) << 1);
+	m_portb |= ((out >> 3) & 1);
+	m_maincpu->set_input_line(INPUT_LINE_IRQ0, !((out >> 4) & 1));
```

Comment:
You can use function style casts rather than using `static_cast` for things like this.  We also have the `bitswap` and `BIT` helpers for shuffling bits.  And what‘s the point of casting `m_irq_in_prio` to an enum when you’re just going to use it as an integer anyway?